sdk/typescript/src/bcs/index.js

```diff
@@ -0,0 +1,663 @@
+     * // struct Coin {
+     * //   value: u64,
+     * //   owner: vector<u8>, // name // Vec<u8> in Rust
+     * //   is_locked: bool,
+     * // }
+     * 
+     * BCS.registerStructType('Coin', {
+     *   value: BCS.U64,
+     *   owner: BCS.STRING,
+     *   is_locked: BCS.BOOL
+     * });
+     * 
+     * // Created in Rust with diem/bcs 
+     * let rust_bcs_str = '80d1b105600000000e4269672057616c6c65742047757900';
+     * 
+     * // Let's encode the value as well
+     * let test_ser = BCS.ser('Coin', {
+     *   owner: 'Big Wallet Guy',
+     *   value: '412412400000',
+     *   is_locked: false,
+     * });
+     * 
+     * console.assert(MoveBCS.util.toHex(test_ser.toArray()) === rust_bcs_str, 'Whoopsie, result mismatch');
+     * 
+     * @param {String} name Name of the type to register.
+     * @param {Object} fields Fields of the struct. Must be in the correct order.
+     * @return {Object} Returns MoveBCS for chaining.
+     */
+    MoveBCS.registerStructType = function registerStructType(name, fields = {}) {
+        if (fields.constructor !== Object) {
+            throw 'Struct description must be an instance of Object type';
+        }
+
+        let struct = Object.freeze(fields); // Make sure the order doesn't get changed
+
+        // IMPORTANT: we need to store canonical order of fields for each registered
+        // struct so we maintain it and allow developers to use any field ordering in
+        // their code (and not cause mismatches based on field order).
+        let canonicalOrder = Object.keys(struct);
+
+        // Make sure all the types in the fields description are already known 
+        // and that all the field types are strings.
+        for (let type of Object.values(struct)) {
+            if (!type.constructor === String) {
+                throw new Error(`Types must be specified as strings, got ${type}`);
+            }
+
+            if (!this.hasType(type)) {
+                throw new Error(`Type ${type} is not registered`);
+            }
+        }
+
+        this.types.set(name, {
+            encode(data, size = 1024) { return this._encodeRaw(new MoveBCS.BcsWriter(size), data); },
+            decode(data) { return this._decodeRaw(new MoveBCS.BcsReader(data)); },
+            
+            _encodeRaw(writer, data) {
+                for (let key of canonicalOrder) {
+                    MoveBCS.types
+                        .get(struct[key])
+                        ._encodeRaw(writer, data[key]);
+                }
+                return writer;
+            },
+            _decodeRaw(reader) {
+                let result = {};
+                for (let key of canonicalOrder) {
+                    result[key] = MoveBCS.types
+                        .get(struct[key])
+                        ._decodeRaw(reader);
+                }
+                return result;
+            }
+        });
+
+        return this;
+    };
+
+    !function registerPrimitives() {
```

Comment:
the `!function name() { ... }()` pattern (though i understand what it does) seems less straightforward than defining the function (`registerPrimitives()`) normally and calling it elsewhere (here, perhaps in a constructor ?)

If there's a worry the function might get re-used when it shouldn't be, perhaps we can use some Typescript features to solve that.